CMakeLists.txt

```diff
@@ -623,6 +623,9 @@ if (BUILD_TESTING)
                 COMMAND
                   find . -name '${test_case_name}.c.o' -exec objcopy --redefine-syms libcrypto.symbols {} \\\;
             )
+            # if we're building tests, we export the prefixed symbols so tests can link to them
+            set_target_properties(${test_case_name} PROPERTIES LINK_FLAGS
+                "-Wl,--whole-archive s2n_libcrypto.a -Wl,--no-whole-archive")
```

Comment:
Aren't we going to end up with duplicate symbols in this case? This will cause each test to _also_ intern libcrypto which should already be happening in the actual s2n artifact.

Reply:
The symbols aren't found without this- the gist mentions they are private; is there a `-fvisibility` or other option needed to make them public? (or an explicit push to re-write the tests w/o calls to libcrypto?)

Reply:
Hmm... I thought all of this was working. In the interning test script it builds all of the tests successfully as is. Do we know what the difference is now?

Reply:
All symbols in AWS-LC have `hidden` visibility by default. Only explicitly selected symbols have `default` visibility.

If a symbol has `hidden` visibility it affects that symbols linkage in an executable object and shared object. In particular, all hidden symbols will have internal/local linkage in those two object types. (see latest System V ABI http://www.sco.com/developers/gabi/latest/ch4.symtab.html#symbol_value).

This means that when the link-editor creates the shared library `libs2n.so`, all symbols in `s2n_libcrypto.a` with hidden visibility will be local symbols. When you subsequently attempt to compile a unit test executable, the link-editor will look in `libs2n.so` for the definition of one of the referenced libcrypto functions, but is unable to find it due to the internal linkage.

I think this is actually what we want and by design! We don't want to export libcrypto symbols from `libs2n.so`.

You can consider the unit test executable as any other application having a dependency on s2n. Interning AWS-LC means they you hide libcrypto inside s2n. Hence, it makes sense that `libs2n.so` does't export libcrypto symbols. 

if a unit test executable has a direct dependency on functions in libcrypto, we should explicitly declare that dependency to the linker and not rely on it being satisfied transitively through the shared library `libs2n.so`. This is also how any "normal" application that depends on s2n would be compiled, I presume.